File: ios/Classes/Accounts/Detail/Assets/Search/Screens/AssetSearchDataSource.swift
// Copyright 2022 Pera Wallet, LDA

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//
//   AssetSearchDataSource.swift

import Foundation
import MacaroonUIKit
import UIKit

final class AssetSearchDataSource: UICollectionViewDiffableDataSource<AssetSearchSection, AssetSearchItem> {
    init(
        _ collectionView: UICollectionView
    ) {
        super.init(collectionView: collectionView) {
            collectionView, indexPath, itemIdentifier in

            switch itemIdentifier {
            case let .asset(item):
                let cell = collectionView.dequeue(AssetPreviewCell.self, at: indexPath)
                cell.bindData(item)
                return cell

            case .empty:
                let cell = collectionView.dequeue(NoContentCell.self, at: indexPath)
                cell.bindData(AssetListSearchNoContentViewModel())
                return cell
            case .noContent:
                let cell = collectionView.dequeue(NoContentCell.self, at: indexPath)
                cell.bindData(AssetListSearchNoContentViewModel(hasBody: false))
                return cell
            }
        }

        supplementaryViewProvider = { collectionView, kind, indexPath in
            guard let section = AssetSearchSection(rawValue: indexPath.section),
                  section == .assets,
                  kind == UICollectionView.elementKindSectionHeader else {
                return nil
            }

            let view = collectionView.dequeueHeader(SingleLineTitleActionHeaderView.self, at: indexPath)
            view.bindData(
                SingleLineTitleActionViewModel(
                    item: SingleLineIconTitleItem(
                        icon: nil,
                        title: .string("accounts-title-assets".localized)
                    )
                )
            )
            return view
        }

        collectionView.register(AssetPreviewCell.self)
        collectionView.register(header: SingleLineTitleActionHeaderView.self)
        collectionView.register(NoContentCell.self)
    }
}

File: android/app/src/main/java/com/algorand/android/assetsearch/domain/pagination/AssetSearchDataSource.kt
/*
 * Copyright 2022 Pera Wallet, LDA
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License
 */

package com.algorand.android.assetsearch.domain.pagination

import com.algorand.android.assetsearch.domain.model.AssetSearchDTO
import com.algorand.android.assetsearch.domain.model.AssetSearchQuery
import com.algorand.android.assetsearch.domain.repository.AssetSearchRepository
import com.algorand.android.models.Pagination
import com.algorand.android.models.Result
import com.algorand.android.utils.PeraPagingSource
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class AssetSearchDataSource(
    private val assetSearchRepository: AssetSearchRepository,
    private val currentQuery: AssetSearchQuery?
) : PeraPagingSource<String, AssetSearchDTO>() {

    override val logTag: String = AssetSearchDataSource::class.java.simpleName

    override suspend fun loadMore(loadUrl: String): LoadResult<String, AssetSearchDTO> {
        return withContext(Dispatchers.IO) {
            try {
                parseResult(getAssetsByUrl(loadUrl))
            } catch (exception: Exception) {
                LoadResult.Error<String, AssetSearchDTO>(exception)
            }
        }
    }

    override suspend fun initializeData(): LoadResult<String, AssetSearchDTO> {
        return parseResult(searchAssets())
    }

    private suspend fun getAssetsByUrl(currentUrlToFetch: String): Result<Pagination<AssetSearchDTO>> {
        return assetSearchRepository.getAssetsByUrl(currentUrlToFetch)
    }

    private suspend fun searchAssets(): Result<Pagination<AssetSearchDTO>> {
        val queryText = currentQuery?.queryText ?: DEFAULT_ASSET_QUERY.queryText
        val hasCollectibles = currentQuery?.hasCollectibles
        return assetSearchRepository.searchAsset(queryText, hasCollectibles)
    }

    private fun parseResult(result: Result<Pagination<AssetSearchDTO>>): LoadResult<String, AssetSearchDTO> {
        return when (result) {
            is Result.Success -> {
                LoadResult.Page(data = result.data.results, prevKey = null, nextKey = result.data.next)
            }
            is Result.Error -> LoadResult.Error<String, AssetSearchDTO>(result.exception)
        }
    }

    companion object {
        val DEFAULT_ASSET_QUERY = AssetSearchQuery.createDefaultQuery()
    }
}
